### DISCUSSION TITLE
Connect to new WIFI

### DISCUSSION THREAD

**Discussion by wilwizard1979** [HASwitchPlate/openHASP#462] [Q&A] [answered]
Hi I built some screens for my Sunton E3248S035C and noticed that when I went on holidays for my beach house I had to reconnect it to a new WIFI.
I didn´t find an easy way to do that, other than re-flashing the firmware and restoring my code.
Telnet does not work without the WIFI connection and I could not find a straightforward Serial connection app (maybe I haven´t tried the correct one).
Is there any easier way to do that?

**Comment by fvanroie** [comment]
Either reflash are use the serial console. There are dozens of good serial apps, or use the [Chrome or Edge web browser](https://nightly.openhasp.com/).

You can also configure the [wifi settings via the serial or telnet console](https://www.openhasp.com/0.6.3/configuration/wifi/#command-line):
 
```
ssid myAccessPointName
pass myWifiPassword
reboot
```

**Reply by wilwizard1979** [reply]
Thanks, maybe I couldn´t precise my question on the right way.
I could fix the WIFI settings re-flashing my display but I was looking for a different way without the need of re-flashing and eventually losing part of your code.
I moved to a vacation house where wifi is different. I could connect my notebook to the new network but I could no longer connect to the display using Telnet for example (since Telnet will need the IP of the display, which is not visible in the new wifi).
Using the Serial Application listed in the document seems to be only useful for re-flashing but not for being using as a serial console terminal which you could send new commands.

**Reply by fvanroie** [reply, marked as answer]
>  Using the Serial Application listed in the document seems to be only useful for re-flashing but not for being using as a serial console terminal which you could send new commands.

You can use any serial terminal application to connect to the serial port and use it as a terminal.
Just select the correct COM port and set the speed to 115200 e.g. in PuTTY:
![image](https://user-images.githubusercontent.com/15969459/231547422-bd76e5a0-4755-4a00-92d5-0208d92f5b20.png)

Then type the commands above.

**Reply by wilwizard1979** [reply]
Thanks a lot @fvanroie. It works with putty.
For others reading this thread, please make sure to match the Speed defined in Putty to the Baudrate of the Debug Settings in OpenHASP.
Before you move to another Wi-Fi environment, ensure to know what is the Baudrate setting you have... ;-)

![image](https://user-images.githubusercontent.com/108979356/232112472-08f7c1fe-2b71-4d9b-9bfd-cf409fa64866.png)

**Comment by DirkO4YB** [comment]
Hello, stepping into this discussion, i would like to change wifi settings (on a Lanbon board) without connecting to a serial connection. Isn't there a way to force the board into the AP mode (like on the first startup), maybe when the given ssid is not found after a (re)boot? Or perhaps by a pushbutton on a settings-page (on the screen in offline-mode)? Greetz, Dirk

**Reply by fvanroie** [reply]
in v0.7.0 you can assign any command to an onscreen button. So you can use that to reset wifi ot factory settings.

**Reply by DirkO4YB** [reply]
Thanks for your quick reply @fvanroie. Do i understand correctly that when i would make ssid and password both empty strings (¨¨) that the system would fall into the AP mode, waiting for new settings?

**Reply by DirkO4YB** [reply]
Doesn't work that way if i put empty strings for ssid and pass via the serial connection, it just loses connection but doesn't go back to factory startup. Is there a command that would force into AP mode? I can't find the right one...

**Reply by fvanroie** [reply]
Try command `config/wifi {"sssid":"","pass":""}`

**Reply by DirkO4YB** [reply]
Thanks again, i will try that. But first i will have to overcome an issue with the webinterface breaking the wifi connection. I need to investigate that first..

**Reply by DirkO4YB** [reply]
Mmmm, i tried to put the command in a button on a page, but the page will not display. Am i doing the syntax wrong?
This is what i thought it should be:

{"id":2,"obj":"btn","x":60,"y":270,"w":60,"h":50,"border_color":"#353535","border_width":2,"radius":15,"bg_color":"#a9a9","text":"Enter","text_color":"white","text_font":16, "action": {"down": "config/wifi{"ssid":"","pass":""}"} }

**Reply by fvanroie** [reply]
The quotes are getting mixed up, I think this should work: `"action": {"down": "config/wifi{'ssid':'','pass':''}"}`

**Reply by DirkO4YB** [reply]
You are right ofcourse, why didn't think of that...
So now the page shows, and when i press the button, the command changes ssid and pass to "" , but after that, the AP is not started to get new values. In stead, the plate keeps retrying STA connection with the following cyclus of messages on the serial port: Unknown error (0); ready; station start; handshake timeout; No Connection... retry xx
This is the same behaviour as i saw when giving the command(s) though the serial connection...

How can we force that the plate goes to AP mode where we can give new credentials (ssid & pass)?

**Reply by wilwizard1979** [reply]
Did you try the command reboot or restart? 
Or later to power off->on?
I know the hardware reset is not practical but just to understand if a reset might be needed?

**Reply by DirkO4YB** [reply]
Yes, i did try both, reboot (via serial) and power off-on. I think a call to the right function (in custom) might do the trick, but i can't figure out which...

**Reply by fvanroie** [reply]
If the SSID is empty then the QR code and AP should show up at the next reboot.
So, you're saying that setting the SSID to an empty string plus a reboot command doesn't work?

What's the output of `http://ipaddress/api/config/wifi/` ?

**Reply by DirkO4YB** [reply]
After some further investigation; when i push the button as discussed before, in the serial monitor the following appears:

[1000D[0KPrompt > [1000D[9C[21:47:54.227][24564/26080  5][47572/47620  1] MSGR: Json text = config/wifi{'ssid':'','pass':''}
[1000D[0KPrompt > [1000D[9C[21:47:54.239][24564/26080  5][47572/47620  1] MSGR: config/wifi{'ssid':'','pass':''}

but further nothing changes; wifi stays connected with the old ssid, also after reboot or power off/on. The output of /api/config/wifi/ confirms that. I noticed that when ssid and pass are set to "" via direct command (via serial) the following feedback appears:

[0KPrompt > [1000D[9C[20:30:42.343][22516/25532 11][44992/45120  1] MSGR: ssid=""
[0KPrompt > ssid ""[1000D[16C[20:30:42.353][22516/25460 11][44992/45120  1] ----: config partition opened
[0KPrompt > ssid ""[1000D[16C[20:30:42.364][22516/24980  9][44992/45120  1] WIFI: {"ssid":"\"\""}

[0KPrompt > ssid[1000D[13C[1000D[0KPrompt > [1000D[9C[20:31:12.096][22516/25532 11][44992/45120  1] MSGR: pass=""
[0KPrompt > pass ""[1000D[16C[20:31:12.106][22516/25460 11][44992/45120  1] ----: config partition opened
[0KPrompt > pass ""[1000D[16C[20:31:12.117][22516/24980  9][44992/45120  1] WIFI: {"pass":"\"\""}
[0KPrompt > pass ""[1000D[16C[20:31:12.133][22516/25460 11][44992/45120  1] ----:     * Wrote pass => "" (2 bytes)

After a reboot the plate actually starts in AP mode and new wifi credentials can be given. The difference is that the direct command triggers the config to be changed, and the command via the pushbutton doesn't. 

I tried to change the mixing of the hyphens to see if that makes a difference, like this: 

"action": {'down': 'config/wifi{"ssid":"","pass":""}' }

Now the feedback on pushing the button is:

[1000D[0KPrompt > [1000D[9C[21:18:53.707][21492/23412  8][47616/47676  1] MSGR: Json text = config/wifi{"ssid":"","pass":""}
[1000D[0KPrompt > [1000D[9C[21:18:53.719][21492/23412  8][47616/47676  1] MSGR: config/wifi{"ssid":"","pass":""}

but the effect is the same; no change of config/wifi.

**Reply by fvanroie** [reply]
There needs to a space between the config/wifi and the json payload: `"action": {"down": "config/wifi {'ssid':'','pass':''}"}`

If you want to use the `ssid` command in the console you just leave the value empty and use `ssid=` (no quotes there).

**Reply by DirkO4YB** [reply]
Confirmed that with the space (and the quotes correct) the plate steps into AP mode (with the barcode screen) after reboot or power off-on. Thanks @fvanroie for the explanations to make it work. Exactly what i was looking for.

**Reply by DirkO4YB** [reply]
Final remark on this, before ssid (and pass) are emptied (""), one must also empty mqtthost (with extra command; in console with mqtt=).
When mqtthost is not empty, the AP is not stable; in console we see:
[1000D[0KPrompt > [1000D[9C[18:11:09.810][  260/ 1096 76][55632/56140  1] WIFI: 5sec mode AP 3
[1000D[0KPrompt > [1000D[9C[18:11:14.811][  140/  420 66][55632/56140  1] WIFI: 5sec mode AP 3
[1000D[0KPrompt > [1000D[9C[18:11:19.811][  140/  444 68][55632/56140  1] WIFI: 5sec mode AP 3
[1000D[0KPrompt > [1000D[9C[18:11:24.490][  260/ 1092 76][55632/56140  1] MQTT: Transport error
[1000D[0KPrompt > [1000D[9C[18:11:24.499][  260/ 1092 76][55632/56140  1] MQTT: Disconnected
[1000D[0KPrompt > [1000D[9C[18:11:24.811][  260/ 1092 76][55632/56140  1] WIFI: 5sec mode AP 3
[1000D[0KPrompt > [1000D[9C[18:11:29.811][  140/  420 66][55632/56140  1] WIFI: 5sec mode AP 3
[1000D[0KPrompt > [1000D[9C[18:11:34.811][  916/ 2496 63][55632/56140  1] WIFI: 5sec mode AP 3
[1000D[0KPrompt > [1000D[9C[18:11:39.512][ 1332/ 3020 55][55632/56140  1] MQTT: Transport error
[1000D[0KPrompt > [1000D[9C[18:11:39.521][ 1332/ 3020 55][55632/56140  1] MQTT: Disconnected

Because the AP is interrupted for a MQTT try-without-succes, the setting form (on 192.168.4.1) will not load.
After clearing mqtthost (and reboot) the console shows:

[1000D[0KPrompt > [1000D[9C[18:14:37.193][17396/19112  8][55632/56140  1] WIFI: 5sec mode AP 3
[1000D[0KPrompt > [1000D[9C[18:14:42.191][16372/18864 13][55632/56140  1] WIFI: 5sec mode AP 3
[1000D[0KPrompt > [1000D[9C[18:14:47.192][16372/19200 14][55632/56140  1] WIFI: 5sec mode AP 3
[1000D[0KPrompt > [1000D[9C[18:14:52.192][16372/19200 14][55632/56140  1] WIFI: 5sec mode AP 3
[1000D[0KPrompt > [1000D[9C[18:14:57.192][16372/19200 14][55632/56140  1] WIFI: 5sec mode AP 3
[1000D[0KPrompt > [1000D[9C[18:15:02.192][16372/19200 14][55632/56140  1] WIFI: 5sec mode AP 3

An uninterrupted AP, and the setting form loads.

**Comment by AdrienV** [comment]
Thanks for this thread, I added this button in my pages.jsonl, it delete current Wifi setting, restart the screen, and show the wifi configuration page on new launch :)

```
{
    "page": 1,
    "id": 1,
    "y": 0,
    "x":0,
    "w": 280,
    "h": 50,
    "obj": "btn",
    "text": "Change Wifi Network",
    "text_font":32,
    "action":{"down": "config/wifi {'ssid':'','pass':''}", "up": "restart"}
}
```
